open_flamingo/train/train_utils.py

```diff
@@ -64,193 +67,314 @@ def train_one_epoch(
     )  # avg time to load one batch of both C4 AND laion (= 1 batch regardless of gradient accum)
     end = time.time()
 
-    # loop through dataloader
-    for num_steps, (batch_laion, batch_mmc4) in tqdm(
-        enumerate(zip(laion_loader, mmc4_loader)),
-        disable=args.rank != 0,
-        total=total_training_steps,
-        initial=(epoch * num_batches_per_epoch),
-    ):
-        data_time_m.update(time.time() - end)
-
-        global_step = num_steps + epoch * num_batches_per_epoch
-
-        #### LAION FORWARD PASS ####
-        images = (
-            batch_laion[0]
-            .to(device_id, dtype=cast_dtype, non_blocking=True)
-            .unsqueeze(1)
-            .unsqueeze(1)
-        )
-
-        input_ids = batch_laion[1][0].to(device_id, dtype=cast_dtype, non_blocking=True)
-        attention_mask = batch_laion[1][1].to(
-            device_id, dtype=cast_dtype, non_blocking=True
-        )
-
-        labels = input_ids.clone()
-        labels[labels == tokenizer.pad_token_id] = -100
-        labels[:, 0] = -100
-        labels[labels == media_token_id] = -100
-        labels.to(device_id)
-
-        with autocast():
-            loss_laion = model(
-                vision_x=images,
-                lang_x=input_ids,
-                attention_mask=attention_mask,
-                labels=labels,
-            )[0]
-        divided_loss_laion = loss_laion / args.gradient_accumulation_steps
-
-        #### C4 FORWARD PASS ####
-        images = (
-            batch_mmc4[0]
-            .to(device_id, dtype=cast_dtype, non_blocking=True)
-            .unsqueeze(2)
-        )
-        input_ids = torch.stack([x[0] for x in batch_mmc4[1]]).squeeze(1)
-        attention_mask = torch.stack([x[1] for x in batch_mmc4[1]]).squeeze(1)
-
-        # NOTE: irena: expected shape of clip_text_input_ids / attention_mask is (N, I, max_seq_len)
-        labels = input_ids.clone()
-        labels[labels == tokenizer.pad_token_id] = -100
-        labels[:, 0] = -100
-
-        for i in range(labels.shape[0]):
-            # remove loss for any token before the first <image> token
-            label_idx = 0
-            while (
-                label_idx < labels.shape[1] and labels[i][label_idx] != media_token_id
-            ):
-                labels[i][label_idx] = -100
-                label_idx += 1
+    if mmc4_loader is not None:
```

Comment:
I do not like the if/else approach. There is a lot of duplicate code that will make it very difficult to maintain this.

Reply:
Can you please refactor this to remove any duplicate code? For example the forward pass from laion, the code for masking out gradients in the embedding, iteration through dataloaders, wandb logging all have a bit of repetition.

Reply:
sure, I wrote some code already, I'll make PR in short.